Google.Api.Gax.Grpc/Rest/RestChannel.cs

```diff
@@ -105,51 +117,97 @@ private async Task AddAuthHeadersAsync(HttpRequestMessage request, RestMethod me
 
-        internal static Status GetStatus(Task<HttpResponseMessage> httpResponseTask) => httpResponseTask.Status switch
+        private static Status GetStatus(Task<ReadHttpResponseMessage> httpResponseTask) => httpResponseTask.Status switch
         {
-            TaskStatus.RanToCompletion => GetStatus(httpResponseTask.Result),
+            TaskStatus.RanToCompletion => httpResponseTask.Result.GetStatus(),
             TaskStatus.Faulted => new Status(StatusCode.Unknown, "HTTP task faulted", httpResponseTask.Exception.InnerException),
             TaskStatus.Canceled => new Status(StatusCode.Cancelled, "Request cancelled"),
             _ => throw new InvalidOperationException("Cannot call GetStatus with an incomplete HTTP call")
         };
 
-        // TODO: What about error details? Are they still available?
-        internal static Status GetStatus(HttpResponseMessage response) =>
-            new Status(ConvertHttpToGrpcStatusCode(response.StatusCode), "");
-
-        private static StatusCode ConvertHttpToGrpcStatusCode(HttpStatusCode httpStatusCode) => httpStatusCode switch
-        {
-            HttpStatusCode.OK => StatusCode.OK,
-            HttpStatusCode.Unauthorized => StatusCode.Unauthenticated,
-            HttpStatusCode.Forbidden => StatusCode.PermissionDenied,
-            HttpStatusCode.BadRequest => StatusCode.InvalidArgument,
-            HttpStatusCode.InternalServerError => StatusCode.Internal,
-            HttpStatusCode.NotFound => StatusCode.NotFound,
-            HttpStatusCode.Conflict => StatusCode.FailedPrecondition,
-            // TODO: Others!
-            _ => StatusCode.Unknown
-        };
-
-        private Metadata GetTrailers(Task<HttpResponseMessage> httpResponseTask)
+        private Metadata GetTrailers(Task<ReadHttpResponseMessage> httpResponseTask)
         {
             if (!httpResponseTask.IsCompleted)
             {
                 throw new InvalidOperationException("Cannot call GetTrailers with an incomplete HTTP call");
             }
-            // We never have any trailers.
-            return new Metadata();
+            return httpResponseTask.Result.GetTrailers();
+        }
+
+        /// <summary>
+        /// In <see cref="AsyncUnaryCall{TResponse}"/> the functions to obtain the TResponse
+        /// and the <see cref="Status"/> of the call are two different functions.
+        /// The function to obtain the response is async, but the function to obtain the
+        /// <see cref="Status"/> is not.
+        /// For being able to surface error details in <see cref="Status"/> we need to be
+        /// able to call <see cref="HttpContent.ReadAsStringAsync"/> which is an async method,
+        /// and thus cannot be done, without blocking, on the sync function that obtains the 
+        /// <see cref="Status"/> in the <see cref="AsyncUnaryCall{TResponse}"/>.
+        /// So we need to make async content reading part of sending the call and not part of
```

Comment:
I think we'd want to remove this sentence when we make it a real PR.

Reply:
I removed the "as it is now" bit. But I'm happy to remove the whole sentence, or even to turn this into a normal comment as a reminder for us.

Reply:
Just realised this is only internal anyway, so we don't need to worry *too* much.